### PR TITLE
Adding impl/test crumbs for Grid

### PR DESCRIPTION
**Pull request by john-science**
## What is the change?

Adding impl/test crumbs for Grid.

## Why is the change being made?

This is part of on-going requirement work.

---

## Checklist

- [X] This PR has only [one purpose or idea](https://terrapower.github.io/armi/developer/tooling.html#one-idea-one-pr).
- [X] [Tests](https://terrapower.github.io/armi/developer/tooling.html#test-it) have been added/updated to verify that the new/changed code works.

- [X] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [X] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [X] The [release notes](https://terrapower.github.io/armi/release/index.html) (location `doc/release/0.X.rst`) are up-to-date with any important changes.
- [X] The [documentation](https://terrapower.github.io/armi/developer/tooling.html#document-it) is still up-to-date in the `doc` folder.
- [X] No [requirements](https://terrapower.github.io/armi/developer/tooling.html#watch-for-requirements) were altered.
- [X] The dependencies are still up-to-date in `pyproject.toml`.

### CODE REVIEW

File: armi/reactor/grids/tests/test_grids.py [hunk 1]
```diff
@@ -308,26 +308,67 @@ def test_getSymmetricIdenticalsThird(self):
...
+        # check 1/3 core
+        self.assertEqual(third.getMinimumRings(2), 2)
+        self.assertEqual(third.getIndicesFromRingAndPos(2, 2), (0, 1))
+        self.assertEqual(third.getPositionsInRing(3), 12)
+        self.assertEqual(third.getSymmetricEquivalents((3, -2)), [(-1, 3), (-2, -1)])
+
+    def test_tipsUpFlatsUp(self):
+        """Test the pointedEndUp attribute of the fromPitch method.
+
+        .. test:: Build a tips-up and a flats-up hexagonal grids.
```
Comment (albeanth) [inline armi/reactor/grids/tests/test_grids.py:356]:
```suggestion
        .. test:: Build a points-up and a flats-up hexagonal grids.
```

File: armi/reactor/grids/tests/test_grids.py [hunk 2]
```diff
@@ -308,26 +308,67 @@ def test_getSymmetricIdenticalsThird(self):
...
+        with self.assertRaises(IndexError):
+            full.getSymmetricEquivalents((3, -2))
+
+        # check 1/3 core
+        self.assertEqual(third.getMinimumRings(2), 2)
+        self.assertEqual(third.getIndicesFromRingAndPos(2, 2), (0, 1))
+        self.assertEqual(third.getPositionsInRing(3), 12)
+        self.assertEqual(third.getSymmetricEquivalents((3, -2)), [(-1, 3), (-2, -1)])
+
+    def test_tipsUpFlatsUp(self):
```
Comment (albeanth) [inline armi/reactor/grids/tests/test_grids.py:353]:
```suggestion
    def test_pointsUpFlatsUp(self):
```

File: armi/reactor/grids/tests/test_grids.py [hunk 3]
```diff
@@ -308,26 +308,67 @@ def test_getSymmetricIdenticalsThird(self):
...
+                geometry.DomainType.THIRD_CORE, geometry.BoundaryType.PERIODIC
+            )
         )
 
+        # check full core
+        self.assertEqual(full.getMinimumRings(2), 2)
+        self.assertEqual(full.getIndicesFromRingAndPos(2, 2), (0, 1))
+        self.assertEqual(full.getPositionsInRing(3), 12)
+        with self.assertRaises(IndexError):
+            full.getSymmetricEquivalents((3, -2))
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/reactor/grids/tests/test_grids.py:345]: We can't "get symmetric equivalents" for a full core.  That's the idea.

For both full and 1/3 core, we have the same (working) helper methods for positions and rings. 

The only difference is you can get symmetric equivalents for 1/3 core, but not full.

File: armi/reactor/grids/structuredgrid.py [hunk 4]
```diff
@@ -290,7 +290,12 @@ def restoreBackup(self):
         self._unitSteps, self._bounds, self._offset = self._backup
 
     def getCoordinates(self, indices, nativeCoords=False) -> numpy.ndarray:
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/reactor/grids/structuredgrid.py:292]: https://github.com/terrapower/armi/blob/2571216afadbd0b773555f3dfc3d77b66840fa8f/armi/reactor/grids/thetarz.py#L92-L98

File: armi/materials/tests/test_materials.py [hunk 5]
```diff
@@ -60,7 +60,16 @@ def test_TD(self):
...
+        """Test the material duplication.
+
+        .. test:: Test the material base class.
+            :id: T_ARMI_MAT_PROPERTIES3
+            :tests: R_ARMI_MAT_PROPERTIES
+
+        .. test:: Materials shall calc mass fracs at init.
+            :id: T_ARMI_MAT_FRACS
+            :tests: R_ARMI_MAT_FRACS
+        """
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/materials/tests/test_materials.py]: Fixed.

File: armi/reactor/components/component.py [hunk 6]
```diff
@@ -260,6 +260,10 @@ def __lt__(self, other):
         True if a circle encompassing this object has a smaller diameter than one encompassing another component.
 
         This allows sorting because the Python sort functions only use this method.
+
+        .. impl:: Order components by there outermost diameter.
+            :id: I_ARMI_COMP_ORDER
+            :implements: R_ARMI_COMP_ORDER
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/reactor/components/component.py]: Fixed.

File: armi/reactor/composites.py [hunk 7]
```diff
@@ -676,6 +676,10 @@ def hasFlags(self, typeID: TypeSpec, exact=False):
             :id: I_ARMI_CMP_FLAG
             :implements: R_ARMI_CMP_FLAG
 
+        .. impl:: Composites have flags
+            :id: I_ARMI_CMP_FLAG
+            :implements: R_ARMI_CMP_FLAG
+
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/reactor/composites.py:678]: Fixed.

File: armi/reactor/components/component.py [hunk 8]
```diff
@@ -260,10 +260,6 @@ def __lt__(self, other):
         True if a circle encompassing this object has a smaller diameter than one encompassing another component.
 
         This allows sorting because the Python sort functions only use this method.
-
-        .. impl:: Order components by there outermost diameter.
-            :id: I_ARMI_COMP_ORDER
-            :implements: R_ARMI_COMP_ORDER
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/reactor/components/component.py]: Fixed.

File: armi/reactor/grids/tests/test_grids.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (albeanth) [reply armi/reactor/grids/tests/test_grids.py:345]: If you set the symmetry
```python
full = grids.HexGrid.fromPitch(1.0)
full.symmetry = str(
    geometry.SymmetryType(
        geometry.DomainType.FULL_CORE, geometry.BoundaryType.NO_SYMMETRY
    )
)
```
then `HexGrid::getSymmetricEquivalents` will return an empty list and no error will get raised. 
https://github.com/terrapower/armi/blob/0a4bad779a5e2abad29653910db1885ea340418e/armi/reactor/grids/hexagonal.py#L325-L326

So I would argue that what has been written doesn't test that `getSymmetricEquivalents` doesn't work for a full core, but that if `self.symmetry` isn't set to anything that the symmetry property can't be evaluated. See the traceback for the index error
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\aalberti\codes\nala_2\framework\armi\reactor\grids\grid.py", line 94, in symmetry
    return geometry.SymmetryType.fromStr(self._symmetry)
  File "C:\Users\aalberti\codes\nala_2\framework\armi\reactor\geometry.py", line 340, in fromStr
    domain = DomainType.fromStr(pieces[0])
IndexError: list index out of range
```

Thoughts?

**Review by albeanth — APPROVED** [review]

**Pull request merged**